File: LeetCode_142/main.cpp
/**
    LeetCode_142
    https://leetcode.com/problems/linked-list-cycle-ii/description/
    MatrixDeity, 2017.
*/

#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* detectCycle(ListNode* head) {
        if (head == nullptr)
            return head;

        ListNode *ptr1 = head, *ptr2 = head;
        while (ptr2 != nullptr && ptr2->next != nullptr) {
            ptr1 = ptr1->next;
            ptr2 = ptr2->next->next;
            if (ptr1 == ptr2)
                break;
        }
        if (ptr2 == nullptr || ptr2->next == nullptr)
            return nullptr;
        ptr1 = head;
        while (ptr1 != ptr2) {
            ptr1 = ptr1->next;
            ptr2 = ptr2->next;
        }
        return ptr1;
    }
};

int main() {
    return 0;
}

File: LeetCode_142/main.go
/**
  LeetCode_142
  https://leetcode.com/problems/linked-list-cycle-ii/
  MatrixDeity, 2023.
*/

package leetcode142

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if fast == slow {
			break
		}
	}

	if fast == nil || fast.Next == nil {
		return nil
	}

	slow = head

	for slow != fast {
		slow = slow.Next
		fast = fast.Next
	}

	return slow
}
